### PR TITLE
[FLINK-1979] Add logistic loss, hinge loss and regularization penalties 

### PR DESCRIPTION
**Pull request by skavulya**
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.
- [ ] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)
- [ ] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added
- [ ] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

…zation

### CODE REVIEW

File: flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/optimization/GradientDescent.scala [hunk 1]
```diff
@@ -272,7 +272,7 @@ abstract class GradientDescent extends IterativeSolver {
   * The regularization function is `1/2 ||w||_2^2` with `w` being the weight vector.
   */
 class GradientDescentL2 extends GradientDescent {
-
+  //TODO(skavulya): Pass regularization penalty as a parameter
```
Comment (tillrohrmann) [inline flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/optimization/GradientDescent.scala]: Is this TODO still valid? If so, can we resolve it?

File: flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/optimization/GradientDescent.scala [hunk 2]
```diff
@@ -287,11 +287,8 @@ class GradientDescentL2 extends GradientDescent {
       regularizationConstant: Double,
       learningRate: Double)
     : Vector = {
-    // add the gradient of the L2 regularization
-    BLAS.axpy(regularizationConstant, weightVector, gradient)
 
-    // update the weights according to the learning rate
-    BLAS.axpy(-learningRate, gradient, weightVector)
+    L2Regularization.takeStep(weightVector, gradient, regularizationConstant,learningRate)
```
Comment (tillrohrmann) [inline flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/optimization/GradientDescent.scala]: whitespace missing between comma and `learningRate`

File: flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/optimization/GradientDescent.scala [hunk 3]
```diff
@@ -306,7 +303,7 @@ object GradientDescentL2 {
   * The regularization function is `||w||_1` with `w` being the weight vector.
   */
 class GradientDescentL1 extends GradientDescent {
-
+  //TODO(skavulya): Pass regularization penalty as a parameter
```
Comment (tillrohrmann) [inline flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/optimization/GradientDescent.scala]: TODOs are always good to be resolved :-)

File: flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/optimization/GradientDescent.scala [hunk 4]
```diff
@@ -321,19 +318,8 @@ class GradientDescentL1 extends GradientDescent {
...
-    val shrinkageVal = regularizationConstant * learningRate
-    var i = 0
-    while (i < weightVector.size) {
-      val wi = weightVector(i)
-      weightVector(i) = scala.math.signum(wi) *
-        scala.math.max(0.0, scala.math.abs(wi) - shrinkageVal)
-      i += 1
-    }
+
+    L1Regularization.takeStep(weightVector, gradient, regularizationConstant,learningRate)
```
Comment (tillrohrmann) [inline flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/optimization/GradientDescent.scala]: whitespace missing

File: flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/optimization/GradientDescent.scala [hunk 5]
```diff
@@ -348,7 +334,7 @@ object GradientDescentL1 {
   * No regularization is applied.
   */
 class SimpleGradientDescent extends GradientDescent {
-
+  //TODO(skavulya): Pass regularization penalty as a parameter
```
Comment (tillrohrmann) [inline flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/optimization/GradientDescent.scala]: TODO

File: flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/optimization/GradientDescent.scala [hunk 6]
```diff
@@ -364,8 +350,7 @@ class SimpleGradientDescent extends GradientDescent {
       learningRate: Double)
     : Vector = {
     // Update the weight vector
-    BLAS.axpy(-learningRate, gradient, weightVector)
-    weightVector
+    NoRegularization.takeStep(weightVector, gradient, regularizationConstant,learningRate)
```
Comment (tillrohrmann) [inline flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/optimization/GradientDescent.scala]: whitespace

File: flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/optimization/LossFunction.scala [hunk 7]
```diff
@@ -23,8 +23,8 @@ import org.apache.flink.ml.math.BLAS
 
 /** Abstract class that implements some of the functionality for common loss functions
   *
-  * A loss function determines the loss term $L(w) of the objective function  $f(w) = L(w) +
-  * \lambda R(w)$ for prediction tasks, the other being regularization, $R(w)$.
+  * A loss function determines the loss term `L(w)` of the objective function  `f(w) = L(w) +
+  * lambda*R(w)` for prediction tasks, the other being regularization, `R(w)`.
```
Comment (tillrohrmann) [inline flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/optimization/LossFunction.scala]: Good catch :-)

File: flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/optimization/PartialLossFunction.scala [hunk 8]
```diff
@@ -47,21 +47,106 @@ object SquaredLoss extends PartialLossFunction {
...
   }
 
   /** Calculates the derivative of the [[PartialLossFunction]]
     *
-    * @param prediction
-    * @param label
-    * @return
+    * @param prediction The predicted value
+    * @param label The true value
+    * @return The derivative of the loss function
```
Comment (tillrohrmann) [inline flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/optimization/PartialLossFunction.scala]: Good code completion, thanks :-)

File: flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/optimization/PartialLossFunction.scala [hunk 9]
```diff
@@ -47,21 +47,106 @@ object SquaredLoss extends PartialLossFunction {
...
     */
   override def derivative(prediction: Double, label: Double): Double = {
     (prediction - label)
   }
 }
+
+/** Logistic loss function which can be used with the [[GenericLossFunction]]
+  *
+  *
+  * The [[LogisticLoss]] function implements `log(1 + exp(prediction*label))`
```
Comment (tillrohrmann) [inline flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/optimization/PartialLossFunction.scala]: A minus is missing in the exponent of e.

**Comment by tillrohrmann** [comment]
Thanks for your contribution @skavulya. The changes look really good :-) I especially like your thorough and clear documentation of the different loss functions. Great work.

I had some minor comments. Once fixed, the PR should be good to be merged :-)

File: flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/optimization/GradientDescent.scala (hunk 1 above)
Comment (thvasilo) [inline flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/optimization/GradientDescent.scala]: I've mentioned this in the previous PR but adding here for completeness:

I'm in favor of adding the regularization penalty as a parameter for the optimizer.

However that would involve changes that perhaps beyond the scope of this PR, 
currently with only SGD available we don't have to worry about the applicability of L1/L2 regularization, but should add a note for when L-BFGS get implemented. 

Depending on how much work @skavulya it would be to make the change here, we can choose to have a separate PR for that or include it here.

Comment (thvasilo) [inline flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/optimization/GradientDescent.scala]: I've mentioned this in the previous PR but adding here for completeness:

I'm in favor of adding the regularization penalty as a parameter for the optimizer.

However that would involve changes that perhaps beyond the scope of this PR, 
currently with only SGD available we don't have to worry about the applicability of L1/L2 regularization, but should add a note for when L-BFGS get implemented. 

Depending on how much work @skavulya it would be to make the change here, we can choose to have a separate PR for that or include it here.

File: flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/optimization/RegularizationPenalty.scala [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+
+    // update the weights according to the learning rate
+    BLAS.axpy(-learningRate, gradient, weightVector)
+
+    weightVector
+  }
+
+  /** Adds regularization to the loss value
+    *
+    * The updated loss is `l + lambda * 1/2*||w||_2^2` where `l` is the old loss,
```
Comment (thvasilo) [inline flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/optimization/RegularizationPenalty.scala]: I would recommend spelling out `loss` here, as in some fonts it's hard tell the letter "l" apart from the number 1,

File: flink-libraries/flink-ml/src/test/scala/org/apache/flink/ml/optimization/RegularizationPenaltyITSuite.scala [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+ */
+
+package org.apache.flink.ml.optimization
+
+import org.apache.flink.ml.math.DenseVector
+import org.apache.flink.test.util.FlinkTestBase
+import org.scalatest.{FlatSpec, Matchers}
+
+
+class RegularizationPenaltyITSuite extends FlatSpec with Matchers with FlinkTestBase {
```
Comment (thvasilo) [inline flink-libraries/flink-ml/src/test/scala/org/apache/flink/ml/optimization/RegularizationPenaltyITSuite.scala]: @tillrohrmann Do tests like these that only test components without an `ExecutionEnvironment` being created get named differently (i.e. no ITSuite in the class name)

File: flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/optimization/GradientDescent.scala (hunk 1 above)
Comment (skavulya) [inline flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/optimization/GradientDescent.scala]: I can add it to this PR since it is only affecting SGD.

File: flink-libraries/flink-ml/src/test/scala/org/apache/flink/ml/optimization/RegularizationPenaltyITSuite.scala (hunk 11 above)
Comment (skavulya) [inline flink-libraries/flink-ml/src/test/scala/org/apache/flink/ml/optimization/RegularizationPenaltyITSuite.scala]: Thanks @tillrohrmann  @thvasilo for all your comments. I'll make the changes for passing the regression penalty as a parameter, fix the formatting issue and update the unit tests

Comment (tillrohrmann) [inline flink-libraries/flink-ml/src/test/scala/org/apache/flink/ml/optimization/RegularizationPenaltyITSuite.scala]: You're right @thvasilo. This would be not an integration test because it tests the regularization component without firing up a Flink cluster and executing an actual job. So we could remove the `ITSuite` from the name.

**Comment by skavulya** [comment]
@tillrohrmann @thvasilo I made the changes you recommended. Please let me know if they look ok.

File: flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/optimization/PartialLossFunction.scala [hunk 12]
```diff
@@ -47,21 +47,106 @@ object SquaredLoss extends PartialLossFunction {
...
+    // based on implementation in scikit-learn
+    // approximately equal and saves the computation of the log
+    if (z > 18) {
+      return math.exp(-z)
+    }
+    else if (z < -18) {
+      return -z
+    }
+
+    math.log(1 + math.exp(-z))
```
Comment (chiwanpark) [inline flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/optimization/PartialLossFunction.scala]: Using `return` is not recommended in Scala. Could you change this like following?

```scala
if (z > 18) {
  math.exp(-z)
} else if (z < -18) {
  -z
} else {
  math.log(1 + math.exp(-z))
}
```

File: flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/optimization/PartialLossFunction.scala [hunk 13]
```diff
@@ -47,21 +47,106 @@ object SquaredLoss extends PartialLossFunction {
...
+    // approximately equal and saves the computation of the log
+    if (z > 18) {
+      return -label * math.exp(-z)
+    }
+    else if (z < -18) {
+      return -label
+    }
+
+    -label/(math.exp(z) + 1)
+  }
```
Comment (chiwanpark) [inline flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/optimization/PartialLossFunction.scala]: As I said above, following is better:

```scala
if (z > 18) {
  -label * math.exp(-z)
} else if (z < -18) {
  -label
} else {
  -label / (math.exp(z) + 1)
}
```

File: flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/optimization/RegularizationPenalty.scala [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+    // Update weight vector with gradient.
+    BLAS.axpy(-learningRate, gradient, weightVector)
+
+    // Apply proximal operator (soft thresholding)
+    val shrinkageVal = regularizationConstant * learningRate
+    var i = 0
+    while (i < weightVector.size) {
+      val wi = weightVector(i)
+      weightVector(i) = scala.math.signum(wi) *
+        scala.math.max(0.0, scala.math.abs(wi) - shrinkageVal)
```
Comment (chiwanpark) [inline flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/optimization/RegularizationPenalty.scala]: We can change `scala.math` to `math`.

File: flink-libraries/flink-ml/src/test/scala/org/apache/flink/ml/optimization/LossFunctionTest.scala [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+
+package org.apache.flink.ml.optimization
+
+import org.apache.flink.ml.common.{LabeledVector, WeightVector}
+import org.apache.flink.ml.math.DenseVector
+import org.scalatest.{Matchers, FlatSpec}
+import org.apache.flink.test.util.FlinkTestBase
+
+
+class LossFunctionTest extends FlatSpec with Matchers with FlinkTestBase {
```
Comment (chiwanpark) [inline flink-libraries/flink-ml/src/test/scala/org/apache/flink/ml/optimization/LossFunctionTest.scala]: We don't need to extend `FlinkTestBase` because `LossFunctionTest` do not use Flink cluster.

File: flink-libraries/flink-ml/src/test/scala/org/apache/flink/ml/optimization/RegularizationPenaltyTest.scala [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+ */
+
+package org.apache.flink.ml.optimization
+
+import org.apache.flink.ml.math.DenseVector
+import org.apache.flink.test.util.FlinkTestBase
+import org.scalatest.{FlatSpec, Matchers}
+
+
+class RegularizationPenaltyTest extends FlatSpec with Matchers with FlinkTestBase {
```
Comment (chiwanpark) [inline flink-libraries/flink-ml/src/test/scala/org/apache/flink/ml/optimization/RegularizationPenaltyTest.scala]: Same as above, we don't need to extend `FlinkTestBase`.

**Comment by chiwanpark** [comment]
Hi @skavulya, I just quickly reviewed your updated PR and left few comments. They are not critical things but It would be better to fix them. Other things are very good.

**Comment by skavulya** [comment]
Thanks @chiwanpark. I made the changes you recommended. Please let me know if it looks ok.

**Comment by skavulya** [comment]
While working on L-BFGS, I realized that I need to remove the gradient descent step from RegularizationPenalty. I'll update the PR soon.

**Comment by chiwanpark** [comment]
Okay, please ping me when the PR is updated.

**Comment by skavulya** [comment]
@chiwanpark Decoupling the gradient descent step is complicated for L1 regularization because we are using the proximal gradient method that applies soft thresholding after executing the gradient descent step. I left the regularization penalty as-is. I am thinking of adding an additional method that adds the regularization penalty to gradient without the gradient descent step but I will do it in the L-BFGS PR instead.

**Comment by skavulya** [comment]
@chiwanpark The PR is ready. Let me know if I need to do anything else.

**Comment by chiwanpark** [comment]
Hi @skavulya, sorry for late response. I've checked the updated PR and looks good to me. I wonder whether the name `RegularizationPenalty` is proper because the class calculates lots of values. But it is trivial thing.

If there is no objection in few days, I'll merge this. Thanks! 👍

**Comment by skavulya** [comment]
hi @chiwanpark sorry, I hadn't checked this PR for a while. I merged the latest master. Do you have any preference on an alternative name for RegularizationPenalty? Would Regularizer work better?

**Comment by mtunique** [comment]
Could this PR be merged ? Maybe this PR blocks to design GLM([FLINK-2013](https://issues.apache.org/jira/browse/FLINK-2013)).

**Comment by tillrohrmann** [comment]
The PR looks good to me. Thanks a lot for your contribution @skavulya. I will merge this PR once the rebased code passes Travis :-)

**Comment by skavulya** [comment]
Thanks @tillrohrmann!
